File: go/kdb/dag/in_memory_commit_dag_test.go
package dag

import (
	"errors"
	"testing"

	"github.com/limidus/kdb/go/kdb/codec"
	"github.com/limidus/kdb/go/kdb/document"
)

func TestAppendAndHeadMovesLinear(t *testing.T) {
	dag, err := NewInMemoryCommitDag("ns")
	if err != nil {
		t.Fatal(err)
	}
	root, err := dag.Head()
	if err != nil {
		t.Fatal(err)
	}
	txID, _ := codec.RandomUUID()
	author, _ := codec.RandomUUID()
	tx := document.Transaction{
		ID: txID, BaseVersion: root, Operations: nil,
		Timestamp: codec.TimestampNow(), AuthorNodeID: author,
	}
	tree := document.EmptyDocumentTree()
	c, err := dag.AppendCommit(tx, root, tree, nil, "")
	if err != nil {
		t.Fatal(err)
	}
	head, err := dag.Head()
	if err != nil {
		t.Fatal(err)
	}
	if c.Hash != head {
		t.Fatal("head did not advance")
	}
	if _, ok := dag.GetCommit(c.Hash); !ok {
		t.Fatal("commit not stored")
	}
}

func TestPutCommitIdempotent(t *testing.T) {
	dag, err := NewInMemoryCommitDag("ns")
	if err != nil {
		t.Fatal(err)
	}
	root, _ := dag.Head()
	txID, _ := codec.RandomUUID()
	author, _ := codec.RandomUUID()
	tx := document.Transaction{
		ID: txID, BaseVersion: root,
		Timestamp: codec.TimestampNow(), AuthorNodeID: author,
	}
	c, err := dag.AppendCommit(tx, root, document.EmptyDocumentTree(), nil, "")
	if err != nil {
		t.Fatal(err)
	}
	if err := dag.PutCommit(c, true); err != nil {
		t.Fatal(err)
	}
	if err := dag.PutCommit(c, true); err != nil {
		t.Fatal(err)
	}
	head, _ := dag.Head()
	if c.Hash != head {
		t.Fatal("head changed")
	}
}

func TestDiffMaps(t *testing.T) {
	dag, err := NewInMemoryCommitDag("ns")
	if err != nil {
		t.Fatal(err)
	}
	id1, _ := codec.RandomUUID()
	id2, _ := codec.RandomUUID()
	id3, _ := codec.RandomUUID()
	h1 := codec.Hash{Bytes: [32]byte{1}}
	h2 := codec.Hash{Bytes: [32]byte{2}}
	h1m := codec.Hash{Bytes: [32]byte{3}}
	h3 := codec.Hash{Bytes: [32]byte{4}}

	root, _ := dag.Head()
	tx1 := newTx(root)
	treeA, err := document.BuildDocumentTree(map[codec.UUID]codec.Hash{id1: h1, id2: h2})
	if err != nil {
		t.Fatal(err)
	}
	ca, err := dag.AppendCommit(tx1, root, treeA, nil, "")
	if err != nil {
		t.Fatal(err)
	}
	tx2 := newTx(ca.Hash)
	treeB, err := document.BuildDocumentTree(map[codec.UUID]codec.Hash{id1: h1m, id3: h3})
	if err != nil {
		t.Fatal(err)
	}
	cb, err := dag.AppendCommit(tx2, ca.Hash, treeB, nil, "")
	if err != nil {
		t.Fatal(err)
	}
	d, err := dag.Diff(ca.Hash, cb.Hash)
	if err != nil {
		t.Fatal(err)
	}
	mod := d.Modified()
	if len(mod) != 1 || mod[0].DocID != id1 {
		t.Fatal("modified")
	}
	added := d.Added()
	if len(added) != 1 || added[0].DocID != id3 {
		t.Fatal("added")
	}
	removed := d.Removed()
	if len(removed) != 1 || removed[0].DocID != id2 {
		t.Fatal("removed")
	}
}

func TestEmptyDiffSameCommit(t *testing.T) {
	dag, err := NewInMemoryCommitDag("ns")
	if err != nil {
		t.Fatal(err)
	}
	h, _ := dag.Head()
	d, err := dag.Diff(h, h)
	if err != nil {
		t.Fatal(err)
	}
	if !d.IsEmpty() {
		t.Fatal("expected empty diff")
	}
}

func TestStubShowsInWalk(t *testing.T) {
	dag, err := NewInMemoryCommitDag("ns")
	if err != nil {
		t.Fatal(err)
	}
	genesis, _ := dag.Head()
	c1, err := dag.AppendCommit(newTx(genesis), genesis, document.EmptyDocumentTree(), nil, "")
	if err != nil {
		t.Fatal(err)
	}
	c2, err := dag.AppendCommit(newTx(c1.Hash), c1.Hash, document.EmptyDocumentTree(), nil, "")
	if err != nil {
		t.Fatal(err)
	}
	if _, err := dag.StubCommit(genesis, "ice://bucket/obj"); err != nil {
		t.Fatal(err)
	}
	w := dag.Walk(c2.Hash, nil, 10)
	found := false
	for _, e := range w {
		if _, ok := e.(StubbedEntry); ok {
			found = true
		}
	}
	if !found {
		t.Fatal("expected stub in walk")
	}
}

func TestCompactionBlockedWhenBranchHeadInSquash(t *testing.T) {
	dag, err := NewInMemoryCommitDag("ns")
	if err != nil {
		t.Fatal(err)
	}
	head, _ := dag.Head()
	_, err = dag.Squash([]codec.Hash{head}, head, document.EmptyDocumentTree(), nil, "")
	var safety *CompactionSafetyError
	if !errors.As(err, &safety) {
		t.Fatalf("expected CompactionSafetyError, got %v", err)
	}
}

func TestMissingParentRejected(t *testing.T) {
	dag, err := NewInMemoryCommitDag("ns")
	if err != nil {
		t.Fatal(err)
	}
	bogus := codec.Hash{Bytes: [32]byte{0x7f}}
	tx, _ := codec.RandomUUID()
	author, _ := codec.RandomUUID()
	bad, err := document.BuildCommit(
		[]codec.Hash{bogus}, "ns", tx, codec.TimestampNow(), author,
		nil, document.EmptyDocumentTree().TreeHash, nil, "",
	)
	if err != nil {
		t.Fatal(err)
	}
	err = dag.PutCommit(bad, true)
	var cons *ConsistencyError
	if !errors.As(err, &cons) {
		t.Fatalf("expected ConsistencyError, got %v", err)
	}
}

func newTx(base codec.Hash) document.Transaction {
	id, _ := codec.RandomUUID()
	author, _ := codec.RandomUUID()
	return document.Transaction{
		ID: id, BaseVersion: base,
		Timestamp: codec.TimestampNow(), AuthorNodeID: author,
	}
}

File: kdb-dag/src/commonTest/kotlin/dev/kdb/dag/InMemoryCommitDagTest.kt
package dev.kdb.dag

import dev.kdb.codec.KdbHash
import dev.kdb.codec.KdbTimestamp
import dev.kdb.codec.KdbUuid
import dev.kdb.document.DocumentTree
import dev.kdb.document.KdbCommit
import dev.kdb.document.KdbTransaction
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class InMemoryCommitDagTest {
    /**
     * Matches Go `dag.NewInMemoryCommitDag("app/t")` genesis. Regenerated
     * after the DocumentTree hash algorithm changed to the trie-based
     * scheme (see the gap-fix note in docs/benchmarks/phases-1-6-summary.md)
     * - both sides were computed independently and agree, which is itself
     * a cross-language interop check for the empty tree's hash.
     */
    @Test
    fun genesisCommitHash_matchesGoForNamespaceAppSlashT() =
        runTest {
            val dag = inMemoryCommitDag("app/t")
            assertEquals(
                "136e831b1bb7a2ed3949e5264b17233117d0868627030f9ccea4a601559ec0ab",
                dag.head().toHex(),
            )
        }

    @Test
    fun tc01_appendAndHeadMovesLinear() =
        runTest {
            val dag = inMemoryCommitDag("ns")
            val root = dag.head()
            val tx =
                KdbTransaction(
                    id = KdbUuid.random(),
                    baseVersion = root,
                    operations = emptyList(),
                    timestamp = KdbTimestamp.now(),
                    authorNodeId = KdbUuid.random(),
                )
            val tree = DocumentTree.EMPTY
            val c = dag.appendCommit(tx, root, tree, schemaHash = null)
            assertEquals(c.hash, dag.head())
            assertNotNull(dag.getCommit(c.hash))
        }

    /**
     * Regression test for a MutableMap.putIfAbsent JVM-only-extension compile bug found while
     * chasing docs/kdb-finish-up-plan.md's 1-K10 (this file is unrelated to that finding, but
     * fixing the compile break here - putCommitLocked's txIndex bookkeeping didn't compile for
     * Kotlin/JS at all - is what made ./gradlew allTests reach every other module in the first
     * place). Verifies the "first commit for a transaction id wins" semantic the JVM-only
     * putIfAbsent used to implement is preserved by its portable containsKey-then-set
     * replacement: appending two different commits that (illegally, but that's exactly the
     * defensive case) share one transaction id must keep resolving that id to the first commit.
     */
    @Test
    fun getCommitByTransactionId_firstCommitForATransactionIdWins() =
        runTest {
            val dag = inMemoryCommitDag("ns")
            val root = dag.head()
            val tx =
                KdbTransaction(
                    id = KdbUuid.random(),
                    baseVersion = root,
                    operations = emptyList(),
                    timestamp = KdbTimestamp.now(),
                    authorNodeId = KdbUuid.random(),
                )
            val tree = DocumentTree.EMPTY
            val first = dag.appendCommit(tx, root, tree, schemaHash = null, message = "first")
            val second = dag.appendCommit(tx, first.hash, tree, schemaHash = null, message = "second")

            assertTrue(first.hash != second.hash, "test setup: the two commits must actually be distinct")
            assertEquals(first.hash, dag.getCommitByTransactionId(tx.id)?.hash, "the first commit for this transaction id must still win")
        }

    @Test
    fun tc02_putCommitIdempotent() =
        runTest {
            val dag = inMemoryCommitDag("ns")
            val root = dag.head()
            val tx =
                KdbTransaction(
                    id = KdbUuid.random(),
                    baseVersion = root,
                    operations = emptyList(),
                    timestamp = KdbTimestamp.now(),
                    authorNodeId = KdbUuid.random(),
                )
            val c = dag.appendCommit(tx, root, DocumentTree.EMPTY, null)
            dag.putCommit(c, requireParents = true)
            dag.putCommit(c, requireParents = true)
            assertEquals(c.hash, dag.head())
        }

    @Test
    fun tc06_diffMaps() =
        runTest {
            val dag = inMemoryCommitDag("ns")
            val id1 = KdbUuid.random()
            val id2 = KdbUuid.random()
            val id3 = KdbUuid.random()
            val h1 = KdbHash.fromBytes(ByteArray(32) { 1 })
            val h2 = KdbHash.fromBytes(ByteArray(32) { 2 })
            val h1m = KdbHash.fromBytes(ByteArray(32) { 3 })
            val h3 = KdbHash.fromBytes(ByteArray(32) { 4 })

            val root = dag.head()
            val t1 =
                KdbTransaction(
                    id = KdbUuid.random(),
                    baseVersion = root,
                    operations = emptyList(),
                    timestamp = KdbTimestamp.now(),
                    authorNodeId = KdbUuid.random(),
                )
            val treeA =
                DocumentTree.build(
                    mapOf(id1 to h1, id2 to h2),
                )
            val ca = dag.appendCommit(t1, root, treeA, null)

            val t2 =
                KdbTransaction(
                    id = KdbUuid.random(),
                    baseVersion = ca.hash,
                    operations = emptyList(),
                    timestamp = KdbTimestamp.now(),
                    authorNodeId = KdbUuid.random(),
                )
            val treeB =
                DocumentTree.build(
                    mapOf(id1 to h1m, id3 to h3),
                )
            val cb = dag.appendCommit(t2, ca.hash, treeB, null)

            val d = dag.diff(ca.hash, cb.hash)
            assertEquals(id1, d.modified.single().docId)
            assertEquals(id3, d.added.single().docId)
            assertEquals(id2, d.removed.single().docId)
        }

    @Test
    fun tc07_emptyDiffSameCommit() =
        runTest {
            val dag = inMemoryCommitDag("ns")
            val h = dag.head()
            assertTrue(dag.diff(h, h).isEmpty)
        }

    @Test
    fun tc09_stubShowsInWalk() =
        runTest {
            val dag = inMemoryCommitDag("ns")
            val genesis = dag.head()
            val tx1 =
                KdbTransaction(
                    id = KdbUuid.random(),
                    baseVersion = genesis,
                    operations = emptyList(),
                    timestamp = KdbTimestamp.now(),
                    authorNodeId = KdbUuid.random(),
                )
            val c1 = dag.appendCommit(tx1, genesis, DocumentTree.EMPTY, null)
            val tx2 =
                KdbTransaction(
                    id = KdbUuid.random(),
                    baseVersion = c1.hash,
                    operations = emptyList(),
                    timestamp = KdbTimestamp.now(),
                    authorNodeId = KdbUuid.random(),
                )
            val c2 = dag.appendCommit(tx2, c1.hash, DocumentTree.EMPTY, null)
            dag.stubCommit(genesis, archiveLocation = "ice://bucket/obj")
            val w = dag.walk(from = c2.hash, limit = 10)
            assertTrue(w.any { it is TraversalEntry.Stubbed })
        }

    @Test
    fun compactionBlockedWhenBranchHeadInSquash() =
        runTest {
            val dag = inMemoryCommitDag("ns")
            val head = dag.head()
            assertFailsWith<CompactionSafetyException> {
                dag.squash(
                    squashHashes = listOf(head),
                    boundary = head,
                    syntheticTree = DocumentTree.EMPTY,
                    syntheticSchemaHash = null,
                )
            }
        }

    @Test
    fun missingParentRejected() =
        runTest {
            val dag = inMemoryCommitDag("ns")
            val bogusParent = KdbHash.fromBytes(ByteArray(32) { 0x7f })
            val bad =
                KdbCommit.build(
                    parentHashes = listOf(bogusParent),
                    namespaceId = "ns",
                    transactionId = KdbUuid.random(),
                    timestamp = KdbTimestamp.now(),
                    authorNodeId = KdbUuid.random(),
                    operations = emptyList(),
                    documentTreeHash = DocumentTree.EMPTY.treeHash,
                    schemaHash = null,
                )
            assertFailsWith<DagConsistencyException> {
                dag.putCommit(bad, requireParents = true)
            }
        }
}
